README

		AMAP v5.4
		Application MAPper

		- a next generation scanning tool

		- by van Hauser and DJ RevMoon / THC <amap-dev@thc.org>

		http://www.thc.org


		NOTE: THIS TOOL IS OUTDATED. USE NMAP.



INTRODUCTION
============

This is a public release of Amap. Amap is a next-generation scanning
tool for pentesters. It attempts to identify applications even if they
are running on a different port than normal. 
It also identifies non-ascii based applications. This is achieved by
sending trigger packets, and looking up the responses in a list of
response strings. 

Without filled databases containing triggers and responses, the tool is
worthless, so I ask you to help us fill the database. How to do this? Well,
whenever a client application connects to a server, some kind of handshake
is exchanged (at least, usually. Syslogd for instance won't say nothing,
and snmpd without the right community string neither). Anyway, amap takes
the first packet sent back and compares it to a list of signature responses.
Really simple, actually. And in reality, it turns out really to be that
simple, at least, for most protocols.

So now, with amap, you can identify that SSL server running on port 3442, 
as well as that Oracle listener on port 23. 

For unknown protocols, you can use amapcrap, which sends random crap to a
udp, tcp or ssl'ed port, to illicit a response, which you can then
put into the appdefs.trig and appdefs.resp files.


INSTALLING
==========

-> please see the INSTALL file for hints and known problems/solutions


TRIGGERS AND RESPONSES
======================

Take a look at the supplied appdefs.trig and appdefs.resp files.
Much will become clear then.


ESSENTIALS
==========

Amap takes nmap -oM output files as input. You can specify a single IP
address and port)s= on the command line, but usually, you'd run it from a nmap
file, thusly:

   # (first "nmap -sS -oM results.nmap -p 1-65535 TARGET" of course)
   # amap -i results.nmap -o results.amap -m

or:

   # amap 127.0.0.1 443

or:

   # amap target 21-6000

Other switches and options can be seen by typing:

    # amap -h


HINTS
=====

use the -1 switch, it makes amap much faster!
use the -b switch to let amap print the banners!
use the -q switch to hide messages about closed ports
use the -v switch to get more information
use the -t and -T switches to adapt to your target servers (timeout options)
use the -W switch to update your fingerprint database online

To have a very fast port scanner and banner grabber (but inefficient
application identification):
  amap -qbp http TARGET 1-65535
and it gives better results than the "native" banner grab mode:
  amap -B -q TARGET 1-65535


CONTRIBUTING
============

Send us the initial packets (sent and received) in tcpdump format for all
wacko, proprietary and obscure applications. Send them to:
amap-dev@thc.org. Please include application name and version.
A web-enabled interface with uploader will become available soon for your
convenience. 
Want to become an amap beta tester? Send an email to amap-subscribe@thc.org


COMMERCIAL COMPANIES/SERVICES
=============================

This tool is not completely GPL! See the LICEN[SC]E.* files.


BUGS
====

None :-)  Send bugreports to amap-dev@thc.org. 


TODO
====

-> see the TODO file


THANKS
======

T1nk, Guido van Rooij, Unicorn, Arhab, Johnny Cyberpunk and many others who
sent us triggers and responses.

Yours,
	van Hauser and DJ RevMoon


CHANGES
Changelog

v5.4	April 2011 (THC public release)
	* Fix for IPv6 introduced in v5.3, doh

v5.3	April 2011 (THC public release)
	! This is a release to fix IPv6 UDP port scanning as no other tool
        ! currently exists to do that.
        ! Beside this, amap is outdated, and nmap should be prefered
        * Fixed UDP port scanning for IPv6
        * Disabled web update
        * added printing of [] brackets around IPv6 addresses (required)
	* Fixed a bug in amapcrap that the trigger would not display correctly

v5.2	September 2005 (THC public release)
	! THIS IS A THC TAX ANNIVERSARY SPECIAL RELEASE ! HAVE FUN !
	* Included patch from ka0ttic@gentoo.org for cleaner gcc compile
	* Added SSL_Pending() to prevent rare locking on SSL ports,
	  thanks to michel(at)arboi.fr.eu.org for reporting
	* Added lots of fingerprints, most from Johnny Cyberpunk / THC - THANKS!

v5.1	June 2005 (THC public release)
	* Big appdefs.resp update. Thanks to all contributors!
	* Finally and forever fixed the --prefix= issue
	* Fixed the web update function for bad inet_pton implementations
	* Added support for nmap files with IPv6 addresses
	* You can scan/check port 0 now (wish from nbach<at>deloitte.dk)
	* Less error prone "make install"

v5.0	February 2005 (THC public release)
	* IPv6 support added, use -6 to activate
	* Made the help screen easier to read
	* No Cygwin detection needed anymore
	* Dump -d option will only print "Unrecognized" response prior those
	  responses which are *really* not unrecognized
	* Uh, appdefs.* files were installed to both, bin/ and etc/, duh
	  finally they are really only in .../etc/
	* Removed the "error - ignored" messages from Makefile
	* Added amap-lib - an open interface so incorporation can be made
	  easily into other programs. See the text file AMAP-LIB-INTERFACE
	* Fixed "lvalue deprecated" warnings during compilation, lets hope
	  amap still compiles on old UNIXes. Please report if not.
	* Fixed prefix issue if "./configure --prefix=/bla" was used

v4.9    February 2005 (THC internal release)
        - internal, see v5.0

v4.8	January 2005 (THC public release)
	* A project web page for amap was set-up: http://www.thc.org/thc-amap
	* Added an Online Update feature for the application fingerprints!
	  Just run it as "amap -W" and there you go! Includes version
	  checking and some nice other features.
	* The application fingerprints of amap will now be installed to
	  /usr/local/etc - this has been asked for sooo long :-)
	* Added many new responses (thanks to finkga@vt.edu,
	  MattT@netforcement.com and others)
	* Fixed a bug in the -q option
	* Note: the license changed, and is now the same as hydra
	* Added autodetection for Cygwin and MacOS/X, enhanced openssl
	  detection

v4.7	October 2004 (THC public release)
	* Fixed a bug in the SSL analyzing function, seems openssl changed
	  behaviour, fingerprinting behing SSL ports works now again
	* Added more triggers and responses (especially Mikhael@nessus.org
	  sent in lots of stuff, thanks a lot)

v4.6	June 2004 (THC public release)
	* Added 9 new responses (thanks to archanab@iplocks.com, Alf, and more)
	* Fixed a small string termination bug (thanks to mjc@bitz.ca)

v4.5	November 2003 (THC public release)
	* Added portability fixes for openbsd (thanks to tony@libpcap.net)
	* Added portability fixes for cygwin (thanks to nexus@patrol.i-way.co.uk)
	* Added 6 responses (thanks to jrdepriest@ftb.com, indole@bu.edu and
	  t.veronesi@inwind.it)

v4.4	September 2003 (unreleased)
	* Added mkdir -p to the Makefile (thanks to mrjoe@host.sk)
	* Added a few responses

v4.3	September 2003 (THC public release)
	* Trigger names are now always printed with unrecognized responses
	* Added 8 new responses (thanks to jai.s@net4india.net,
	  haugh@cs.ucdavis.edu, ngregoire@exaprobe.com, bcalder01@yahoo.com
	  and file13@budweiser.com)
	* Fixed a off-by-one overflow (which would not fuck up anything :-)
	  was found by z33d (thanks to z33d@isec.pl)
	* Added --prefix option to configure script, and honoring PREFIX
	* Enhanced ssl library searching
	* Added PCRE_DOTALL to the pcre regex definition to enhance
	  response identifications
	* Fixed typos :-) (thanks to guys from the CCC Camp presentation)

v4.2	August 2003 (THC public release)
	* Added 10 new responses (thanks to jrdepriest@ftb.com,
	  adarth@gazeta.pl, larry@thorn.net and jc@thc.org)
	* Enhanced again the max size for an nmap line, lets hope 64k are
	  enough now! (thanks to kapkebap@gmx.de)
	* Due a bug, amap would only check the last host line in nmap files
	* Fixed a compiler warning (thanks to yonatan@xpert.com)
	* Trigger name is now displayed in verbose mode by unrecognized
	  responses
        - 4.1 - THC beta release -
	* 28 new response ids and triggers (thanks to jrdepriest@ftb.com,
	  madhat@unspecific.com, k.gavrilenko@arhont.com and yclin@ustc.edu)
	* It was possible to define ports to be scanned > 65535, fixed
	  (found by harl@marsmenschen.com)
	* If appdefs.* files are in MS-DOS file format, this is not a problem
	  any more :-) no need for dos2unix etc. if you received appdefs.*
	  files via M$-Outlook

v4.0	July 2003 (THC public release)
	! This is the first public release of amap after its complete rewrite !
	! If you would like to be an amap beta tester, subscribe yourself to
	  our amap mailing list! send an email to: amap-subscribe@thc.org
	! What is new from the last public version (2.7):
	  + TCP connection reuse for RPC identification
	  + Banner grab mode, Portscan mode (-B, -P)
	  + for response identification (appdefs.resp):
		  * response strings are now real perl regular expressions
		  * can hit only on a defined trigger if wished so
		  * can have a minimum and maximum length set on the reply data
		  * can require the ip protocol (tcp or udp)
	  + Put as many ports on the command line as you like and ranges are
	    supported too! :-)
	  + much faster
	  + more reliable
	  + bug fixes, better platform support

amap.c
/* AMAP - application mapper Copyright (c) 2003 van Hauser and DJ.RevMoon
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.    
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *    
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

// INCLUDES //
#include "amap-inc.h"
#include "amap.h"
#include "amap-lib.h"

int amap6 = 0;

// HELP //
void help(char *prg) {
  printf("%s v%s (c) %s by %s <%s> %s\n", AMAP_PROGRAM, AMAP_VERSION, AMAP_YEAR, AMAP_AUTHOR, AMAP_EMAIL, AMAP_RESOURCE);
  printf("Syntax: %s [-A|-B|-P|-W] [-1buSRHUdqv] [[-m] -o <file>] [-D <file>] [-t/-T sec] [-c cons] [-C retries] [-p proto] [-i <file>] [target port [port] ...]\n", prg);
  printf("Modes:\n");
  printf("  -A         Map applications: send triggers and analyse responses (default)\n");
  printf("  -B         Just grab banners, do not send triggers\n");
  printf("  -P         No banner or application stuff - be a (full connect) port scanner\n");
//  printf("  -W         Web Update - online update the application fingerprint database!\n");
  printf("Options:\n");
  printf("  -1         Only send triggers to a port until 1st identification. Speeeeed!\n");
#ifdef AF_INET6
  if (amap6)
    printf("  -4         Use IPv4 instead of IPv6\n");
  else
    printf("  -6         Use IPv6 instead of IPv4\n");
#endif
  printf("  -b         Print ascii banner of responses\n");
  printf("  -i FILE    Nmap machine readable outputfile to read ports from\n");
  printf("  -u         Ports specified on commandline are UDP (default is TCP)\n");
#ifdef OPENSSL
  printf("  -R / -S    Do NOT identify RPC / SSL services\n");
#else
  printf("  -R         Do NOT identify RPC service\n");
#endif
  printf("  -H         Do NOT send application triggers marked as potentially harmful\n");
  printf("  -U         Do NOT dump unrecognised responses (better for scripting)\n");
  printf("  -d         Dump all responses\n");
  printf("  -v         Verbose mode, use twice (or more!) for debug (not recommended :-)\n");
  printf("  -q         Do not report closed ports, and do not print them as unidentified\n");
  printf("  -o FILE [-m] Write output to file FILE, -m creates machine readable output\n");
  printf("  -c CONS    Amount of parallel connections to make (default %d, max %d)\n", AMAP_DEFAULT_TASKS, AMAP_MAX_TASKS);
  printf("  -C RETRIES Number of reconnects on connect timeouts (see -T) (default %d)\n", AMAP_MAX_CONNECT_RETRIES);
  printf("  -T SEC     Connect timeout on connection attempts in seconds (default %d)\n", AMAP_CONNECT_TIME);
  printf("  -t SEC     Response wait timeout in seconds (default %d)\n", AMAP_RESPONSE_TIME);
  printf("  -p PROTO   Only send triggers for this protocol (e.g. ftp)\n");
//  printf("  -D FILE    Read from Definitions FILE[.trig|.resp|.rpc] instead of default\n");
//  printf("  -h         Print this shit\n");
  printf("  TARGET PORT   The target address and port(s) to scan (additional to -i)\n");
  printf("%s is a tool to identify application protocols on target ports.\n", AMAP_PROGRAM);
#ifndef OPENSSL
  printf("Note: this version was NOT compiled with SSL support!\n");
#endif
  printf("Usage hint: Options \"-bqv\" are recommended, add \"-1\" for fast/rush checks.\n");
  exit(-1);
}


// MAIN //
int main(int argc, char *argv[]) {
  // VARIABLES //
  amap_struct_options   *opt;
  int newargc;
  char *newargv[argc];
  int  i = 0;

  // INITIALISATION //
  opt = amap_main_init();

  if (strstr(argv[0], "amap6") != NULL) {
    amap6 = 1;
    opt->ipv6 = 1;
  }

  // GETOPT //
  if (argc < 2 || strncmp(argv[1], "-?", 2) == 0 || strncmp(argv[1], "--h", 3) == 0)
    help(argv[0]);
  while ((i = getopt(argc, argv, "146SRHUbuvdhmi:T:c:C:p:o:D:t:qABPW")) >= 0) {
    switch (i) {
    case 'A': // defines in the future that we want default AMAP mode
      break;
    case '1':
      opt->one_is_enough = 1;
      break;
    case '4':
      opt->ipv6 = 0;
      break;
    case '6':
      opt->ipv6 = 1;
#ifndef AF_INET6
      amap_error("No IPv6 support found on your system");
#endif
      break;
    case 'd':
      opt->dump_all = 1;
      break;
    case 'v':
      opt->verbose++;
      break;
    case 'S':
      opt->do_scan_ssl = 0;
      break;
    case 'R':
      opt->do_scan_rpc = 0;
      break;
    case 'b':
      opt->banner = 1;
      break;
    case 'm':
      opt->machine_readable = 1;
      break;
    case 'H':
      opt->harmful = 0;
      break;
    case 'U':
      opt->dump_unidentified = 0;
      break;
    case 'c':
      opt->tasks = atoi(optarg);
      break;
    case 'C':
      opt->max_connect_retries = atoi(optarg);
      break;
    case 'i':
      opt->file_nmap = optarg;
      break;
    case 'u':
      opt->cmd_proto = AMAP_PROTO_UDP;
      break;
    case 'p':
      opt->only_send_trigger = optarg;
      break;
    case 'o':
      opt->file_log = optarg;
      break;
    case 'q':
      opt->quiet = 1;
      break;
    case 'D':
      opt->filename = optarg;
      break;
    case 't':
      opt->timeout_response = atoi(optarg);
      break;
    case 'T':
      opt->timeout_connect = atoi(optarg);
      break;
    case 'B':
      opt->banner_only = 1;
      opt->do_scan_ssl = 0;
      opt->do_scan_rpc = 0;
      opt->one_is_enough = 1;
      break;
    case 'P':
      opt->portscanner = 1;
      opt->do_scan_ssl = 0;
      opt->do_scan_rpc = 0;
      opt->one_is_enough = 1;
      opt->timeout_response--;
      break;
    case 'W':
      fprintf(stderr, "Error: web update is not available anymore\n");
      exit(-1);
      opt->update = 1;
      break;
    case 'h':
      help(argv[0]);
      break;
    default:
      fprintf(stderr, "Error: unknown option -%c\n", i);
      help(argv[0]);
    }
  }

  // VARIABLES VERIFICATION //
  if (opt->update && argc != 2 && opt->filename == NULL)
    amap_warn("amap takes no other commandline options when in -W online update mode except -D");
  if ((optind + 2 > argc) && (opt->file_nmap == NULL) && (opt->update == 0))
    help(argv[0]);
  if (opt->file_log == NULL && opt->machine_readable)
    amap_error("option -m set, but no logfile defined (-o)");
  if (opt->tasks < 1 || opt->tasks > AMAP_MAX_TASKS)
    amap_error("the connect task option (-c) must be between 1 and %d", AMAP_MAX_TASKS);
  if (opt->timeout_connect < 1 || opt->timeout_connect > 240)
    amap_error("the connect timeout option (-T) must be between 1 and 240, its counted in seconds!");
  if (opt->timeout_response < 1 || opt->timeout_response > 240)
    amap_error("the response timeout option (-t) must be between 1 and 240, its counted in seconds!");

  newargc = argc - optind;
  for (i = optind; i < argc; i++)
    newargv[i - optind] = argv[i];
  newargv[newargc] = NULL;

  return amap_main(opt, newargc, newargv);
}


amap.h
#ifndef _AMAP_H

/* AMAP - Application MAPper Copyright (c) 2003-2005 van Hauser and DJ RevMoon
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.    
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *    
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#define AMAP_PROGRAM   "amap"
#define AMAP_VERSION   "5.4"
#define AMAP_YEAR      "2011"
#define AMAP_AUTHOR    "van Hauser"
#define AMAP_EMAIL     "vh@thc.org"
#define AMAP_RESOURCE  "www.thc.org/thc-amap"

#ifndef AMAP_PREFIX
 #ifdef PREFIX
  #warning "PREFIX definition found, installing to this prefix directory location"
  #define AMAP_PREFIX         PREFIX
 #else
  #define AMAP_PREFIX         "/usr/local"
 #endif
#endif

#define AMAP_BUFSIZE		1024	// standard buffer size
#define AMAP_BUFSIZE_BIG	65536   // big standard buffer size
#define AMAP_REGEX_OPTIONS	( PCRE_MULTILINE | PCRE_CASELESS | PCRE_DOTALL )

/* web update feature */
#define AMAP_WEBBUFLEN 1024
#define AMAP_MAXTOKENLEN 64

/* connection and task definitions */
#define AMAP_MAX_CONNECT_RETRIES	3	// connect() retries
#define AMAP_CONNECT_TIME	5	// seconds to wait for connect
#define AMAP_RESPONSE_TIME      5	// seconds to wait for response
#define AMAP_MAX_TASKS		256	// maximum parallel tasks
#define AMAP_DEFAULT_TASKS	32	// default parallel tasks
#define AMAP_MAX_ID_LENGTH	32
#define AMAP_UFO		"unidentified"

/* file definitions */
#define AMAP_DEFAULT_FILENAME	"appdefs"	// default filename
#define AMAP_FILETYPE_RESPONSES	".resp"	// default extension
#define AMAP_FILETYPE_TRIGGERS	".trig"	// default extension
#define AMAP_FILETYPE_RPC	".rpc"	// default extension

/* scan modes */
#define AMAP_SCANMODE_DEFAULT	1
#define AMAP_SCANMODE_SSL	2
#define AMAP_SCANMODE_RPC	3

/* ip protocols */
#define AMAP_PROTO_TCP	1
#define AMAP_PROTO_UDP  2
#define AMAP_PROTO_BOTH 3

/* connect states */
#define AMAP_CONNECT_NULL        0
#define AMAP_CONNECT_INPROGRESS  1
#define AMAP_CONNECT_READY       2
#define AMAP_CONNECT_ACTIVE      3
#define AMAP_CONNECT_REUSABLE    4
#define AMAP_CONNECT_RETRY       5

/* all the important structures */
typedef struct {
  char  *only_send_trigger;
  FILE  *logfile;
  int   tasks;
  unsigned char  timeout_connect;
  unsigned char  timeout_response;
  char  max_connect_retries;
  char  do_scan_ssl;
  char  do_scan_rpc;
  char  verbose;
  char  quiet;
  char  banner;
  char  banner_only;
  char  portscanner;
  char  update;
  char  machine_readable;
  char  harmful;
  char  one_is_enough;
  char  dump_unidentified;
  char  dump_all;
  char  ipv6;
  /* for lib package moved here */
  char  *file_nmap;
  char  *file_log;
  char  *filename;
  int   cmd_proto;
} amap_struct_options;

typedef struct {
  unsigned short int port;
  struct amap_struct_portlist *next;
} amap_struct_portlist;

typedef struct {
  char *id;
  amap_struct_portlist *ports;
  char ip_prot;
  char harmful;
  char *trigger;
  int  trigger_length;
  struct amap_struct_triggers *next;
} amap_struct_triggers;

typedef struct {
  char *trigger;
  struct amap_struct_triggerptr *next;
} amap_struct_triggerptr;

typedef struct {
  char *id;
  amap_struct_triggerptr *triggerptr;
  char ip_prot;
  int min_length;
  int max_length;
  pcre *pattern;
  pcre_extra *hints;
  struct amap_struct_responses *next;
} amap_struct_responses;

typedef struct {
  char *id;
  struct amap_struct_identifications *next;
} amap_struct_identifications;

typedef struct {
  unsigned short int port;
  char ip_prot;
  char ssl;
  char rpc;
  char skip;
  int unknown_response_length;
  char *unknown_response;
  amap_struct_identifications *ids;
  struct amap_struct_ports *next;
} amap_struct_ports;

typedef struct {
  char *target;
  amap_struct_ports *ports;
  struct amap_struct_targets *next;
} amap_struct_targets;

typedef struct {
  int running;
  int tasks;
  char scanmode;
} amap_struct_scaninfo;

typedef struct {
  char active;
  char ssl_enabled;
  char retry;
  unsigned char response[AMAP_BUFSIZE + 1];
  int socket;
  int response_length;
  int sockaddr_len;
  time_t timer;
  struct sockaddr *sockaddr;
#ifdef OPENSSL
  SSL *ssl_socket;
#endif
  amap_struct_targets *target;
  amap_struct_ports *port;
  amap_struct_triggers *trigger;
} amap_struct_coms;

#define _AMAP_H
#endif
